src/protocols/bdx/BdxMessages.h

```diff
@@ -79,46 +79,60 @@ enum RangeControlFlags : uint8_t
+ * @brief
+ *   Interface for defining methods that apply to all BDX messages.
  */
-struct TransferInit
+struct BdxMessage
 {
     /**
      * @brief
-     *  Write the message fields to a buffer using the provided BufBound.
+     *  Parse data from an PacketBuffer into a BdxMessage struct.
      *
-     *  It is up to the caller to use BufBound::Fit() to verify that the write was
-     *  successful. This method will also not check for correctness or completeness for
-     *  any of the fields - it is the caller's responsibility to ensure that the fields
-     *  have been filled out adequately.
+     *  Note that this may store pointers that point into the passed PacketBuffer,
+     *  so it is essential that the underlying PacketBuffer is not modified until after this
+     *  struct is no longer needed.
      *
-     * @param aBuffer A BufBound object that will be used to write the message
+     * @param[in] aBuffer A PacketBufferHandle with a refernce to the PacketBuffer containing the data.
+     *
+     * @return CHIP_ERROR Return an error if the message format is invalid and/or can't be parsed
      */
-    BufBound & WriteToBuffer(BufBound & aBuffer) const;
+    CHECK_RETURN_VALUE
+    CHIP_ERROR Parse(System::PacketBufferHandle aBuffer) { return DerivedParse(std::move(aBuffer)); }
 
     /**
      * @brief
-     *  Parse data from an PacketBuffer into a struct instance.
-     *
-     *  Note that this struct will store pointers that point into the passed PacketBuffer,
-     *  so it is essential that the PacketBuffer is not modified until after this
-     *  struct is no longer needed.
+     *  Write the message fields to a buffer using the provided BufBound.
      *
-     * @param[in] aBuffer Pointer to a PacketBuffer containing the data.
+     *  It is up to the caller to use BufBound::Fit() to verify that the write was
+     *  successful. This method will also not check for correctness or completeness for
+     *  any of the fields - it is the caller's responsibility to ensure that the fields
+     *  align with BDX specifications.
      *
-     * @return CHIP_ERROR Any error that occurs when trying to read the message
+     * @param aBuffer A BufBound object that will be used to write the message
      */
-    CHECK_RETURN_VALUE
-    CHIP_ERROR Parse(System::PacketBufferHandle aBuffer);
+    BufBound & WriteToBuffer(BufBound & aBuffer) const { return DerivedWriteToBuffer(aBuffer); }
 
     /**
      * @brief
-     *  Returns the size of buffer needed to write this message.
+     *  Returns the size of buffer needed to write the message.
      */
-    size_t MessageSize() const;
+    virtual size_t MessageSize() const { return DerivedMessageSize(); }
 
+    virtual ~BdxMessage() = default;
+
+private:
+    virtual CHIP_ERROR DerivedParse(System::PacketBufferHandle aBuffer) = 0;
+    virtual BufBound & DerivedWriteToBuffer(BufBound & aBuffer) const   = 0;
+    virtual size_t DerivedMessageSize() const                           = 0;
```

Comment:
Why "derived" for the privates? Why not have MessageSize() be abstract, WriteToBuffer() be abstract, etc.

Virtual private methods is very very weird. 

I suggest you fix this as a follow-up, since it's merely an implementation detail and your public interface remains the same.

I strongly recommend just expressing an abstract public API and override in each impl, and just remove this private section.